### PR TITLE
Collection lazy implementation

### PR DESCRIPTION
**Pull request by ikwattro**
In accordance with #113 . 

Would fix the need to call the getters before being able to add to a collection.

### CODE REVIEW

File: tests/Integration/MovieDatasetTest.php [hunk 1]
```diff
@@ -179,17 +179,22 @@ public function testRelationshipReferenceCanBeReAddedAndRemoved()
...
         $person = $this->em->getRepository(Person::class)->findOneBy(['name' => 'Tom Hanks']);
         $movie = new Movie('Super Movie');
         $person->getMovies()->add($movie);
         $movie->getActors()->add($person);
+        $person->getMovies()->first();
         $this->em->flush();
         $this->assertGraphExist('(p:Person {name:"Tom Hanks"})-[:ACTED_IN]->(m:Movie {title:"Super Movie"})');
         $person->getMovies()->removeElement($movie);
         $movie->getActors()->removeElement($person);
+        $start = microtime(true);
```
Comment (Nyholm) [inline tests/Integration/MovieDatasetTest.php]: Why is this needed?

File: src/Proxy/LazyCollection.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    protected $node;
+
+    protected $object;
+
+    protected $initialized = false;
+
+    protected $initializing = false;
+
+    protected $added = [];
```
Comment (Nyholm) [inline src/Proxy/LazyCollection.php]: Are these properties needed to be protected? We should consider make them private.

Comment (ikwattro) [inline src/Proxy/LazyCollection.php]: Agree.

File: tests/Integration/MovieDatasetTest.php (hunk 1 above)
Comment (ikwattro) [inline tests/Integration/MovieDatasetTest.php]: debug left.

**Comment by ikwattro** [comment]
@JoeThielen @theUm @sts85 @Nyholm 

As you might now, since the RC, when loading an entity from the database, its Relationship properties being collections would be null at first instance on a proxy, forcing the user to call first the getter in order to inititialize the collection.

This PR removes this constraint by having a LazyCollection backed by a Collection object.

No need to trigger the getter first anymore.

The relationships will not be loaded from the db if : 

* count() is called on the collection, rather it will send once a getDegree to the db
* add($element) doesn't load the collection anymore, so you can add relationships without having to load all the relationships

This PR goes into rc3, some additional improvements will come once I have more feedback on the RC's.

Ideas for improvements are as in Doctrine Extra Lazy : 

* containsElement() will not load the relationships
* adding slice() in order to paginate efficiently all the relationships

I will release RC3 this weekend.

Cheers,

Chris

**Comment by ikwattro** [comment]
Fixes #113

**Comment by Nyholm** [comment]
Awesome. Great work!

**Comment by JoeThielen** [comment]
This seems to be working in my testing.  Thanks!

**Comment by ikwattro** [comment]
@JoeThielen Cool to hear

**Comment by sts85** [comment]
Did not run into any problems on my project as well. Great work!
